### PR TITLE
crypto: clean-up impossible errors from `blake2b` and `PublicKeyWithHash`

### PR DESCRIPTION
**Pull request by lierdakil**
Since we're doing the spring clean-up in #62, might as well do this.

This PR removes impossible errors from return types in `blake2b_*` functions with fixed hash size, `merkle_tree`, and conversion from public key to public key hashes.

~~Doesn't touch `PublicKeyWithHash` interface, which includes an `Error`, now essentially unused, in the interest of compatibility.~~

### CODE REVIEW

File: tezos-encoding/src/nom.rs [hunk 1]
```diff
@@ -584,8 +584,7 @@ where
 {
     move |input| {
         let (rest, result) = parser(input)?;
-        let hash = crypto::blake2b::digest_256(&input[..input.len() - rest.len()])
-            .map_err(|e| nom::Err::Failure(NomError::hash_error(input, e)))?;
```
Comment (emturner) [inline tezos-encoding/src/nom.rs:588]: does this make `hash_error` unused too?

File: crypto/src/hash.rs [hunk 2]
```diff
@@ -556,41 +549,41 @@ pk_with_hash!(PublicKeyP256, ContractTz3Hash);
...
         let hash = blake2b::digest_160(&source.0);
-        let typed_hash = Self::from_bytes(&hash)?;
+        let typed_hash = Self::from_bytes(&hash).unwrap();
         Ok(typed_hash)
     }
 }
 
 impl TryFrom<PublicKeyBls> for ContractTz4Hash {
-    type Error = TryFromPKError;
+    type Error = ::std::convert::Infallible;
```
Comment (emturner) [inline crypto/src/hash.rs]: we could remove the `Result` type altogether from `PublicKeyWithHash` no?

File: crypto/src/hash.rs [hunk 3]
```diff
@@ -540,6 +540,13 @@ macro_rules! pk_with_hash {
                 Ok(typed_hash)
             }
         }
+
+        impl From<$pk> for $pkh {
+            fn from(source: $pk) -> Self {
+                let hash = blake2b::digest_160(&source.0);
+                Self::from_bytes(&hash).unwrap()
```
Comment (emturner) [inline crypto/src/hash.rs:547]:
```
source.pk_hash().unwrap()
```

**Comment by emturner** [comment]
Thanks for the PR! couple of suggestions, but otherwise LGTM

File: tezos-encoding/src/nom.rs (hunk 1 above)
Comment (lierdakil) [reply tezos-encoding/src/nom.rs:588]: Good catch! Removed in 7935bacb3e6bf24e562fb1bf24cb548c3652ad1b

File: crypto/src/hash.rs (hunk 2 above)
Comment (lierdakil) [reply crypto/src/hash.rs]: I was a bit concerned about changing the trait interface wrt downstream compatibility, but arguably this isn't that big of a concern. Done in f105e32d21ef6ae79432cbe6299b983ea75377a3

File: crypto/src/hash.rs (hunk 3 above)
Comment (lierdakil) [reply crypto/src/hash.rs]: Done. I have amended the commit, see 33b9179e30b26027236cd825ba97386568160c60.

File: crypto/src/hash.rs (hunk 2 above)
Comment (emturner) [reply crypto/src/hash.rs]: thanks! will be a few things needing to be updated downstream anyway - easier to do it all in one go 😄

**Review by emturner — APPROVED** [review]

**Pull request merged**